drivers/platform/stm32/gpio.c

```diff
@@ -0,0 +1,320 @@
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "util.h"
+#include "gpio.h"
+#include "gpio_extra.h"
+
+/**
+ * @brief stm32 platform specific GPIO platform ops structure
+ */
+const struct gpio_platform_ops stm32_gpio_platform_ops = {
+	.gpio_ops_get = &stm32_gpio_get,
+	.gpio_ops_get_optional = &stm32_gpio_get_optional,
+	.gpio_ops_remove = &stm32_gpio_remove,
+	.gpio_ops_direction_input = &stm32_gpio_direction_input,
+	.gpio_ops_direction_output = &stm32_gpio_direction_output,
+	.gpio_ops_get_direction = &stm32_gpio_get_direction,
+	.gpio_ops_set_value = &stm32_gpio_set_value,
+	.gpio_ops_get_value = &stm32_gpio_get_value,
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Prepare the GPIO decriptor.
+ * @param desc - The GPIO descriptor.
+ * @param param - The structure that contains the GPIO parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t _gpio_init(struct gpio_desc *desc,
+		   const struct gpio_init_param *param)
+{
+	int32_t ret = 0;
+	struct stm32_gpio_desc *extra = desc->extra;
+	struct stm32_gpio_desc *pextra = param->extra;
+	GPIO_InitTypeDef gis;
+
+	if (!param)
+		return -EINVAL;
+
+	/* enable gpio port in RCC */
+	if (pextra->port == GPIOA)
```

Comment:
nitpick: why not using a switch?

Reply:
switch doesnt work for pointers unless you cast, it looks ugly with casts all over
GPIOA is a pointer